src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using System.Threading;
+
+namespace Npgsql.Util
+{
+    /// <summary>
+    /// Internal struct wrapping a <see cref="CancellationTokenSource"/> for timeouts.
+    /// </summary>
+    /// <remarks>
+    /// Since there's no way to reset a <see cref="CancellationTokenSource"/> once it was cancelled,
+    /// we need to make sure that an existing cancellation token source hasn't been cancelled,
+    /// every time we start it (see https://github.com/dotnet/runtime/issues/4694).
+    /// </remarks>
+    class TimeoutCancellationTokenSourceWrapper : IDisposable
+    {
+        public TimeSpan Timeout { get; set; }
```

Comment:
We may want to do a Debug.Assert that the timeout doesn't get changed between Start/Stop

Reply:
I don't know how I'd implement this without the additional complexity of adding an `_isRunning` field which seems a bit much for a simple `Debug.Assert()`.
Also, changing `Timeout` between Start/Stop is neither unintended nor dangerous.
We simply pick up the new value upon the next call to `Start()` and that's it.

The most dangerous thing one can do with this wrapper is to keep a reference to an old `CancellationToken` after calling `Start()` again, but I don't know how I could prevent that.
You can make this mistake with the raw `CancellationTokenSource` too but it would be more obvious in that case.

Reply:
>  Also, changing Timeout between Start/Stop is neither unintended nor dangerous.

Maybe, I just don't think we'd ever need to change the timeout for the next usage before the current usage is completed (via Stop), and so if that ever happened, it's very likely it would be a bug. Yeah, I was thinking of an `_isRunning` field under `#if DEBUG`. Not critical.

Reply:
> I was thinking of an _isRunning field under #if DEBUG.

After all this is definitely not a bad Idea.
I'll add that ASAP.